pkg/modifycontroller/controller.go

```diff
@@ -0,0 +1,378 @@
+}
+
+func (ctrl *modifyController) patchPersistentVolume(oldPV, newPV *v1.PersistentVolume) (*v1.PersistentVolume, error) {
+	patchBytes, err := util.GetPatchData(oldPV, newPV)
+	if err != nil {
+		return nil, fmt.Errorf("can't update capacity of PV %s as generate path data failed: %v", newPV.Name, err)
+	}
+	updatedPV, updateErr := ctrl.kubeClient.CoreV1().PersistentVolumes().Patch(context.TODO(), newPV.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+	if updateErr != nil {
+		return nil, fmt.Errorf("update capacity of PV %s failed: %v", newPV.Name, updateErr)
+	}
+	err = ctrl.volumes.Update(updatedPV)
+	if err != nil {
+		return nil, fmt.Errorf("error updating PV %s in local cache: %v", newPV.Name, err)
+	}
+	return updatedPV, nil
+}
+
+func getObjectKey(obj interface{}) (string, error) {
+	if unknown, ok := obj.(cache.DeletedFinalStateUnknown); ok && unknown.Obj != nil {
+		obj = unknown.Obj
+	}
+	objKey, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get key from object")
+		return "", err
+	}
+	return objKey, nil
+}
+
+// modifyPVC modifies the PVC and PV based on VAC
+func (ctrl *modifyController) modifyPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) error {
+	var err error
+	if isFirstTimeModifyVolumeWithPVC(pvc, pv) {
+		// If it is first time adding a vac, always validate and then call modify volume
+		_, _, err, _ = ctrl.validateVACAndModifyVolumeWithTarget(pvc, pv)
+	} else {
+		_, _, err, _ = ctrl.modify(pvc, pv)
+	}
+	return err
+}
+
+func isFirstTimeModifyVolumeWithPVC(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) bool {
+	if pv.Spec.VolumeAttributesClassName == nil && pvc.Spec.VolumeAttributesClassName != nil {
+		return true
+	}
+	return false
+}
+
+// Run starts the controller.
+func (ctrl *modifyController) Run(
+	workers int, ctx context.Context) {
+	defer ctrl.claimQueue.ShutDown()
+
+	klog.InfoS("Starting external resizer for modify volume", "controller", ctrl.name)
+	defer klog.InfoS("Shutting down external resizer", "controller", ctrl.name)
+
+	stopCh := ctx.Done()
+	informersSyncd := []cache.InformerSynced{ctrl.pvSynced, ctrl.pvcSynced}
+
+	// Only WaitForCacheSync for VAC if feature gate is enabled
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
+		informersSyncd = append(informersSyncd, ctrl.vacSynced)
+	}
+
+	if !cache.WaitForCacheSync(stopCh, informersSyncd...) {
+		klog.ErrorS(nil, "Cannot sync pod, pv, pvc or vac caches")
+		return
+	}
+
+	// Cache all the InProgress/Infeasible PVCs as Uncertain for ModifyVolume
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeAttributesClass) {
+		err := ctrl.initUncertainPVCs()
+		if err != nil {
+			klog.ErrorS(err, "Failed to initialize uncertain pvcs")
+		}
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.syncPVCs, 0, stopCh)
+	}
+
+	<-stopCh
+}
+
+// syncPVCs is the main worker.
+func (ctrl *modifyController) syncPVCs() {
```

Comment:
I'm confused on the func name of syncPVC**s** and syncPVC

```suggestion
func (ctrl *modifyController) sync() {
```

Reply:
The naming without an entity can be confusing too though.